File: extension/android/executorch_android/src/main/java/org/pytorch/executorch/EValue.java
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree.
 */

package org.pytorch.executorch;

import java.nio.ByteBuffer;
import java.util.Arrays;
import java.util.Locale;
import org.pytorch.executorch.annotations.Experimental;

/**
 * Java representation of an ExecuTorch value, which is implemented as tagged union that can be one
 * of the supported types: https://pytorch.org/docs/stable/jit.html#types .
 *
 * <p>Calling {@code toX} methods for inappropriate types will throw {@link IllegalStateException}.
 *
 * <p>{@code EValue} objects are constructed with {@code EValue.from(value)}, {@code
 * EValue.tupleFrom(value1, value2, ...)}, {@code EValue.listFrom(value1, value2, ...)}, or one of
 * the {@code dict} methods, depending on the key type.
 *
 * <p>Data is retrieved from {@code EValue} objects with the {@code toX()} methods. Note that {@code
 * str}-type EValues must be extracted with {@link #toStr()}, rather than {@link #toString()}.
 *
 * <p>{@code EValue} objects may retain references to objects passed into their constructors, and
 * may return references to their internal state from {@code toX()}.
 *
 * <p>Warning: These APIs are experimental and subject to change without notice
 */
@Experimental
public class EValue {
  private static final int TYPE_CODE_NONE = 0;

  private static final int TYPE_CODE_TENSOR = 1;
  private static final int TYPE_CODE_STRING = 2;
  private static final int TYPE_CODE_DOUBLE = 3;
  private static final int TYPE_CODE_INT = 4;
  private static final int TYPE_CODE_BOOL = 5;

  private String[] TYPE_NAMES = {
    "None", "Tensor", "String", "Double", "Int", "Bool",
  };

  final int mTypeCode;
  Object mData;

  private EValue(int typeCode) {
    this.mTypeCode = typeCode;
  }

  public boolean isNone() {
    return TYPE_CODE_NONE == this.mTypeCode;
  }

  
  public boolean isTensor() {
    return TYPE_CODE_TENSOR == this.mTypeCode;
  }

  
  public boolean isBool() {
    return TYPE_CODE_BOOL == this.mTypeCode;
  }

  
  public boolean isInt() {
    return TYPE_CODE_INT == this.mTypeCode;
  }

  
  public boolean isDouble() {
    return TYPE_CODE_DOUBLE == this.mTypeCode;
  }

  
  public boolean isString() {
    return TYPE_CODE_STRING == this.mTypeCode;
  }

  /** Creates a new {@code EValue} of type {@code Optional} that contains no value. */
  
  public static EValue optionalNone() {
    return new EValue(TYPE_CODE_NONE);
  }

  /** Creates a new {@code EValue} of type {@code Tensor}. */
  
  public static EValue from(Tensor tensor) {
    final EValue iv = new EValue(TYPE_CODE_TENSOR);
    iv.mData = tensor;
    return iv;
  }

  /** Creates a new {@code EValue} of type {@code bool}. */
  
  public static EValue from(boolean value) {
    final EValue iv = new EValue(TYPE_CODE_BOOL);
    iv.mData = value;
    return iv;
  }

  /** Creates a new {@code EValue} of type {@code int}. */
  
  public static EValue from(long value) {
    final EValue iv = new EValue(TYPE_CODE_INT);
    iv.mData = value;
    return iv;
  }

  /** Creates a new {@code EValue} of type {@code double}. */
  
  public static EValue from(double value) {
    final EValue iv = new EValue(TYPE_CODE_DOUBLE);
    iv.mData = value;
    return iv;
  }

  /** Creates a new {@code EValue} of type {@code str}. */
  
  public static EValue from(String value) {
    final EValue iv = new EValue(TYPE_CODE_STRING);
    iv.mData = value;
    return iv;
  }

  
  public Tensor toTensor() {
    preconditionType(TYPE_CODE_TENSOR, mTypeCode);
    return (Tensor) mData;
  }

  
  public boolean toBool() {
    preconditionType(TYPE_CODE_BOOL, mTypeCode);
    return (boolean) mData;
  }

  
  public long toInt() {
    preconditionType(TYPE_CODE_INT, mTypeCode);
    return (long) mData;
  }

  
  public double toDouble() {
    preconditionType(TYPE_CODE_DOUBLE, mTypeCode);
    return (double) mData;
  }

  
  public String toStr() {
    preconditionType(TYPE_CODE_STRING, mTypeCode);
    return (String) mData;
  }

  private void preconditionType(int typeCodeExpected, int typeCode) {
    if (typeCode != typeCodeExpected) {
      throw new IllegalStateException(
          String.format(
              Locale.US,
              "Expected EValue type %s, actual type %s",
              getTypeName(typeCodeExpected),
              getTypeName(typeCode)));
    }
  }

  private String getTypeName(int typeCode) {
    return typeCode >= 0 && typeCode < TYPE_NAMES.length ? TYPE_NAMES[typeCode] : "Unknown";
  }

  /**
   * Serializes an {@code EValue} into a byte array. Note: This method is experimental and subject
   * to change without notice.
   *
   * @return The serialized byte array.
   */
  public byte[] toByteArray() {
    if (isNone()) {
      return ByteBuffer.allocate(1).put((byte) TYPE_CODE_NONE).array();
    } else if (isTensor()) {
      Tensor t = toTensor();
      byte[] tByteArray = t.toByteArray();
      return ByteBuffer.allocate(1 + tByteArray.length)
          .put((byte) TYPE_CODE_TENSOR)
          .put(tByteArray)
          .array();
    } else if (isBool()) {
      return ByteBuffer.allocate(2)
          .put((byte) TYPE_CODE_BOOL)
          .put((byte) (toBool() ? 1 : 0))
          .array();
    } else if (isInt()) {
      return ByteBuffer.allocate(9).put((byte) TYPE_CODE_INT).putLong(toInt()).array();
    } else if (isDouble()) {
      return ByteBuffer.allocate(9).put((byte) TYPE_CODE_DOUBLE).putDouble(toDouble()).array();
    } else if (isString()) {
      return ByteBuffer.allocate(1 + toString().length())
          .put((byte) TYPE_CODE_STRING)
          .put(toString().getBytes())
          .array();
    } else {
      throw new IllegalArgumentException("Unknown Tensor dtype");
    }
  }

  /**
   * Deserializes an {@code EValue} from a byte[]. Note: This method is experimental and subject to
   * change without notice.
   *
   * @param bytes The byte array to deserialize from.
   * @return The deserialized {@code EValue}.
   */
  public static EValue fromByteArray(byte[] bytes) {
    ByteBuffer buffer = ByteBuffer.wrap(bytes);
    if (buffer == null) {
      throw new IllegalArgumentException("buffer cannot be null");
    }
    if (!buffer.hasRemaining()) {
      throw new IllegalArgumentException("invalid buffer");
    }
    int typeCode = buffer.get();
    switch (typeCode) {
      case TYPE_CODE_NONE:
        return new EValue(TYPE_CODE_NONE);
      case TYPE_CODE_TENSOR:
        byte[] bufferArray = buffer.array();
        return from(Tensor.fromByteArray(Arrays.copyOfRange(bufferArray, 1, bufferArray.length)));
      case TYPE_CODE_STRING:
        throw new IllegalArgumentException("TYPE_CODE_STRING is not supported");
      case TYPE_CODE_DOUBLE:
        return from(buffer.getDouble());
      case TYPE_CODE_INT:
        return from(buffer.getLong());
      case TYPE_CODE_BOOL:
        return from(buffer.get() != 0);
    }
    throw new IllegalArgumentException("invalid type code: " + typeCode);
  }
}

File: extension/java/src/main/java/org/pytorch/executorch/EValue.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree.
 */

package org.pytorch.executorch

import com.facebook.jni.annotations.DoNotStrip
import java.nio.ByteBuffer
import java.nio.charset.StandardCharsets
import java.util.Arrays
import java.util.Locale
import org.pytorch.executorch.annotations.Experimental

/**
 * Java representation of an ExecuTorch value, which is implemented as tagged union that can be one
 * of the supported types: https://pytorch.org/docs/stable/jit.html#types .
 *
 * Calling `toX` methods for inappropriate types will throw [IllegalStateException].
 *
 * `EValue` objects are constructed with `EValue.from(value)`, depending on the value type.
 *
 * Data is retrieved from `EValue` objects with the `toX()` methods. Note that `str`-type EValues
 * must be extracted with [toStr], rather than [toString].
 *
 * `EValue` objects may retain references to objects passed into their constructors, and may return
 * references to their internal state from `toX()`.
 *
 * Warning: These APIs are experimental and subject to change without notice
 */
@Experimental
@DoNotStrip
class EValue
@DoNotStrip
private constructor(
    // JNI reads this field by name via GetFieldID("mTypeCode")
    @JvmField @DoNotStrip val mTypeCode: Int
) {

  // JNI accesses this field by name via GetFieldID("mData"), requires @JvmField for direct field
  // access
  @JvmField @DoNotStrip var mData: Any? = null

  private val typeNames = arrayOf("None", "Tensor", "String", "Double", "Int", "Bool")

  val isNone: Boolean
    @DoNotStrip get() = TYPE_CODE_NONE == mTypeCode

  val isTensor: Boolean
    @DoNotStrip get() = TYPE_CODE_TENSOR == mTypeCode

  val isBool: Boolean
    @DoNotStrip get() = TYPE_CODE_BOOL == mTypeCode

  val isInt: Boolean
    @DoNotStrip get() = TYPE_CODE_INT == mTypeCode

  val isDouble: Boolean
    @DoNotStrip get() = TYPE_CODE_DOUBLE == mTypeCode

  val isString: Boolean
    @DoNotStrip get() = TYPE_CODE_STRING == mTypeCode

  @DoNotStrip
  fun toTensor(): Tensor {
    preconditionType(TYPE_CODE_TENSOR, mTypeCode)
    return mData as? Tensor ?: throw IllegalStateException("EValue data is null or not a Tensor")
  }

  @DoNotStrip
  fun toBool(): Boolean {
    preconditionType(TYPE_CODE_BOOL, mTypeCode)
    return mData as? Boolean ?: throw IllegalStateException("EValue data is null or not a Boolean")
  }

  @DoNotStrip
  fun toInt(): Long {
    preconditionType(TYPE_CODE_INT, mTypeCode)
    return mData as? Long ?: throw IllegalStateException("EValue data is null or not a Long")
  }

  @DoNotStrip
  fun toDouble(): Double {
    preconditionType(TYPE_CODE_DOUBLE, mTypeCode)
    return mData as? Double ?: throw IllegalStateException("EValue data is null or not a Double")
  }

  @DoNotStrip
  fun toStr(): String {
    preconditionType(TYPE_CODE_STRING, mTypeCode)
    return mData as? String ?: throw IllegalStateException("EValue data is null or not a String")
  }

  private fun preconditionType(typeCodeExpected: Int, typeCode: Int) {
    if (typeCode != typeCodeExpected) {
      throw IllegalStateException(
          String.format(
              Locale.US,
              "Expected EValue type %s, actual type %s",
              getTypeName(typeCodeExpected),
              getTypeName(typeCode),
          )
      )
    }
  }

  private fun getTypeName(typeCode: Int): String =
      if (typeCode in typeNames.indices) typeNames[typeCode] else "Unknown"

  /**
   * Serializes an `EValue` into a byte array. Note: This method is experimental and subject to
   * change without notice.
   */
  fun toByteArray(): ByteArray =
      when {
        isNone -> ByteBuffer.allocate(1).put(TYPE_CODE_NONE.toByte()).array()
        isTensor -> {
          val tByteArray = toTensor().toByteArray()
          ByteBuffer.allocate(1 + tByteArray.size)
              .put(TYPE_CODE_TENSOR.toByte())
              .put(tByteArray)
              .array()
        }
        isBool ->
            ByteBuffer.allocate(2)
                .put(TYPE_CODE_BOOL.toByte())
                .put(if (toBool()) 1.toByte() else 0.toByte())
                .array()
        isInt -> ByteBuffer.allocate(9).put(TYPE_CODE_INT.toByte()).putLong(toInt()).array()
        isDouble ->
            ByteBuffer.allocate(9).put(TYPE_CODE_DOUBLE.toByte()).putDouble(toDouble()).array()
        isString -> {
          val strBytes = toStr().toByteArray(StandardCharsets.UTF_8)
          ByteBuffer.allocate(1 + 4 + strBytes.size)
              .put(TYPE_CODE_STRING.toByte())
              .putInt(strBytes.size)
              .put(strBytes)
              .array()
        }
        else -> throw IllegalArgumentException("Unknown EValue type code: $mTypeCode")
      }

  companion object {
    private const val TYPE_CODE_NONE = 0
    private const val TYPE_CODE_TENSOR = 1
    private const val TYPE_CODE_STRING = 2
    private const val TYPE_CODE_DOUBLE = 3
    private const val TYPE_CODE_INT = 4
    private const val TYPE_CODE_BOOL = 5

    /** Creates a new `EValue` of type `Optional` that contains no value. */
    @DoNotStrip @JvmStatic fun optionalNone(): EValue = EValue(TYPE_CODE_NONE)

    /** Creates a new `EValue` of type `Tensor`. */
    @DoNotStrip
    @JvmStatic
    fun from(tensor: Tensor): EValue = EValue(TYPE_CODE_TENSOR).also { it.mData = tensor }

    /** Creates a new `EValue` of type `bool`. */
    @DoNotStrip
    @JvmStatic
    fun from(value: Boolean): EValue = EValue(TYPE_CODE_BOOL).also { it.mData = value }

    /** Creates a new `EValue` of type `int`. */
    @DoNotStrip
    @JvmStatic
    fun from(value: Long): EValue = EValue(TYPE_CODE_INT).also { it.mData = value }

    /** Creates a new `EValue` of type `double`. */
    @DoNotStrip
    @JvmStatic
    fun from(value: Double): EValue = EValue(TYPE_CODE_DOUBLE).also { it.mData = value }

    /** Creates a new `EValue` of type `str`. */
    @DoNotStrip
    @JvmStatic
    fun from(value: String): EValue = EValue(TYPE_CODE_STRING).also { it.mData = value }

    /**
     * Deserializes an `EValue` from a byte[]. Note: This method is experimental and subject to
     * change without notice.
     */
    @JvmStatic
    fun fromByteArray(bytes: ByteArray): EValue {
      val buffer = ByteBuffer.wrap(bytes)
      require(buffer.hasRemaining()) { "invalid buffer" }
      return when (val typeCode = buffer.get().toInt()) {
        TYPE_CODE_NONE -> EValue(TYPE_CODE_NONE)
        TYPE_CODE_TENSOR -> {
          val bufferArray = buffer.array()
          from(Tensor.fromByteArray(Arrays.copyOfRange(bufferArray, 1, bufferArray.size)))
        }
        TYPE_CODE_STRING -> {
          val strLen = buffer.getInt()
          val strBytes = ByteArray(strLen)
          buffer.get(strBytes)
          from(String(strBytes, StandardCharsets.UTF_8))
        }
        TYPE_CODE_DOUBLE -> from(buffer.getDouble())
        TYPE_CODE_INT -> from(buffer.getLong())
        TYPE_CODE_BOOL -> from(buffer.get().toInt() != 0)
        else -> throw IllegalArgumentException("invalid type code: $typeCode")
      }
    }
  }
}
